torch/csrc/autograd/engine.cpp

```diff
@@ -452,10 +452,7 @@ static void validate_outputs(const edge_list& edges, variable_list& grads, const
       grads[i] = at::sum_to(std::move(grads[i]), metadata.shape());
     }
     if (!is_compatible_type(metadata.type(), grads[i].type())) {
-      std::stringstream ss;
-      ss << "invalid gradient at index " << i << " - expected type ";
-      ss << metadata.type() << " but got " << grads[i].type();
-      AT_ERROR(format_error(ss.str()));
+      grads[i] = grads[i].to(metadata.type());
```

Comment:
@gchanan we want to allow integer -> floating point for things like `int_tensor * float_tensor` where the gradient of float_tensor is the value of int_tensor.

I think it's OK to disallow dense -> sparse conversion.

Reply:
@colesbury my thought was that wouldn't come up in practice because you'd always be multiplying with the grad_output, which is going to be float.  But maybe I'm missing something?

Reply:
@gchanan hmmm... yeah I forgot about that

Reply:
I removed the sparse conversion, but I don't think I understand what dtype conversions  we need to preserve and/or check.

Can we get an integer gradient type? I can't create a tensor that requires_grad and has an non-floating dtype.

Where do we multiply by the grad_output?